migrator/migrations/m_181_to_m_182_remove_default_scope_manager_role/migration_impl.go

```diff
@@ -0,0 +1,327 @@
+	if useRoleCheckErr != nil {
+		canRemoveRole = false
+		err = multierror.Append(err, useRoleCheckErr)
+	}
+	if usesScopeManagerRole {
+		addRoleErr := addDeprecatedScopeManagerRole(ctx, roleStorage)
+		if addRoleErr != nil {
+			canRemoveRole = false
+			err = multierror.Append(err, addRoleErr)
+		} else {
+			tokenUpdateErr := updateScopeManagerRoleReferencesInAPITokens(ctx, apiTokenStorage)
+			if tokenUpdateErr != nil {
+				canRemoveRole = false
+				err = multierror.Append(err, tokenUpdateErr)
+			}
+			groupUpdateErr := updateScopeManagerRoleReferencesInGroups(ctx, groupStorage)
+			if groupUpdateErr != nil {
+				canRemoveRole = false
+				err = multierror.Append(err, tokenUpdateErr)
+			}
+		}
+	}
+	if canRemoveRole {
+		roleRemoveErr := removeDefaultScopeManagerRole(ctx, roleStorage)
+		if roleRemoveErr != nil {
+			canRemovePermissionSet = false
+			err = multierror.Append(err, roleRemoveErr)
+		}
+	} else {
+		canRemovePermissionSet = false
+	}
+	// Check for permission set references
+	usesScopeManagerPermissionSet, usePermissionSetCheckErr := isScopeManagerPermissionSetReferenced(ctx, roleStorage)
+	if usePermissionSetCheckErr != nil {
+		canRemovePermissionSet = false
+		err = multierror.Append(err, usePermissionSetCheckErr)
+	}
+	if usesScopeManagerPermissionSet {
+		// The Role -> PermissionSet reference is using the PermissionSet ID which will be left untouched.
+		// Therefore, there is no need to update the PermissionSet reference in the roles.
+		// The permission set Name and Description will be updated.
+		updatePermissionSetErr := updateScopeManagerPermissionSet(ctx, permissionSetStorage)
+		if updatePermissionSetErr != nil {
+			err = multierror.Append(err, updatePermissionSetErr)
+		}
+	} else {
+		if canRemovePermissionSet {
+			deletePermissionSetErr := removeDefaultScopeManagerPermissionSet(ctx, permissionSetStorage)
+			if deletePermissionSetErr != nil {
+				err = multierror.Append(err, deletePermissionSetErr)
+			}
+		}
+	}
+	return err.ErrorOrNil()
+}
+
+func apiTokenHasScopeManagerRole(obj *storage.TokenMetadata) bool {
+	tokenRoles := obj.GetRoles()
+	for _, roleName := range tokenRoles {
+		if roleName == ScopeManagerRoleName {
+			return true
+		}
+	}
+	return false
+}
+
+func groupHasScopeManagerRole(obj *storage.Group) bool {
+	return obj.GetRoleName() == ScopeManagerRoleName
+}
+
+func isScopeManagerRoleReferenced(ctx context.Context, apiTokenStorage apiTokenStore.Store, groupStorage groupStore.Store) (bool, error) {
+	roleReferenceFound := false
+	var err *multierror.Error
+	tokenWalkErr := apiTokenStorage.Walk(ctx, func(obj *storage.TokenMetadata) error {
+		if roleReferenceFound {
+			return nil
+		}
+		hasScopeManagerRole := apiTokenHasScopeManagerRole(obj)
+		if hasScopeManagerRole {
+			roleReferenceFound = true
+		}
+		return nil
+	})
+	if tokenWalkErr != nil {
+		err = multierror.Append(err, tokenWalkErr)
+	}
+	if roleReferenceFound {
```

Comment:
Why do we ignore any potential errors we received during `tokenWalkErr` here?

Reply:
The point here is to check if the default `Scope Manager` role is referenced by any token.
The walk function does not return errors. Potential errors would be on the table walk side, which is likely a DB error.
If a reference was found, the relevant information is `there is a reference`.
If the error was transient, the subsequent walk for role reference update might succeed.
If the error is not a transient one, the migration will fail at the next DB interaction anyway.